File: packages/typescript/tests/instanced-scene.test.ts
import { describe, it, expect } from 'vitest';
import { buildSceneFromParsed, type ParsedDefinition } from '../src/model';
import { buildInstancedSceneFromParsed } from '../src/instanced';
import { GeometryBuilder } from '../src/geometry';
import {
  flattenInstancedScene,
  flattenBakedScene,
  compareTrianglesInOrder,
  instancedBufferBytes,
} from './helpers/flatten-instanced';
import {
  panelDefinition,
  panelWithHoleDefinition,
  texturedMaterial,
  plainMaterial,
  makeParsed,
  translation,
  scaleMatrix,
  rotationZ,
  repeatedComponentScene,
} from './helpers/instanced-fixtures';

/**
 * The instanced scene builder must describe exactly the geometry the baked
 * builder describes - just stored once per definition instead of once per
 * placement. Every test here either proves that equivalence or proves the
 * sharing/variant rules that make the sharing safe.
 */

describe('buildInstancedScene: resource sharing', () => {
  it('emits ONE mesh resource for a definition placed many times', () => {
    const parsed = repeatedComponentScene(50, 4);
    const scene = buildInstancedSceneFromParsed(parsed);

    expect(scene.meshResources.length).toBe(1);
    expect(scene.sceneHierarchy.children.length).toBe(50);
    // every placement points at the same resource
    const ids = new Set(scene.sceneHierarchy.children.map((n) => n.meshResourceId));
    expect(ids.size).toBe(1);
    expect(ids.has(scene.meshResources[0].id)).toBe(true);
  });

  it('keeps mesh-resource buffer bytes flat as instance count grows', () => {
    const sizes = [1, 10, 100].map((n) => instancedBufferBytes(buildInstancedSceneFromParsed(repeatedComponentScene(n, 4))));

    // Identical, not merely sub-linear: the geometry is stored exactly once.
    expect(sizes[1]).toBe(sizes[0]);
    expect(sizes[2]).toBe(sizes[0]);
  });

  it('grows the BAKED path with instance count, which is the problem being solved', () => {
    const bakedPrims = [1, 10, 100].map(
      (n) => buildSceneFromParsed(repeatedComponentScene(n, 4)).glbPrimitives.length
    );
    expect(bakedPrims[1]).toBeGreaterThan(bakedPrims[0]);
    expect(bakedPrims[2]).toBeGreaterThan(bakedPrims[1]);

    // ...while the instanced resource count stays at one.
    const instanced = [1, 10, 100].map(
      (n) => buildInstancedSceneFromParsed(repeatedComponentScene(n, 4)).meshResources.length
    );
    expect(instanced).toEqual([1, 1, 1]);
  });

  it('assigns stable, deterministic resource ids across repeated builds', () => {
    const idsOf = () =>
      buildInstancedSceneFromParsed(repeatedComponentScene(5, 4)).meshResources.map((r) => r.id);
    expect(idsOf()).toEqual(idsOf());
    expect(idsOf()).toEqual(['mesh_0']);
  });
});

describe('buildInstancedScene: material variants in resource identity', () => {
  /** The same definition placed twice, painted per instance. */
  const twoPaintedInstances = (matIdA: number | null, matIdB: number | null) =>
    makeParsed({
      defs: new Map<number | string, ParsedDefinition>([[1, panelDefinition()]]),
      rootInstances: [
        { refIdx: 1, name: 'A', matrix: translation(0, 0, 0), materialId: matIdA },
        { refIdx: 1, name: 'B', matrix: translation(50, 0, 0), materialId: matIdB },
      ],
      materials: new Map([
        ['Red', plainMaterial('Red', { r: 200, g: 10, b: 10, a: 255 })],
        ['Blue', plainMaterial('Blue', { r: 10, g: 10, b: 200, a: 255 })],
      ]),
      materialIdToName: new Map([
        [1, 'Red'],
        [2, 'Blue'],
      ]),
    });

  it('shares ONE resource when the effective inherited material matches', () => {
    const scene = buildInstancedSceneFromParsed(twoPaintedInstances(1, 1));
    expect(scene.meshResources.length).toBe(1);
    const [a, b] = scene.sceneHierarchy.children;
    expect(a.meshResourceId).toBe(b.meshResourceId);
  });

  it('splits into SEPARATE resource variants for different inherited materials', () => {
    const scene = buildInstancedSceneFromParsed(twoPaintedInstances(1, 2));
    expect(scene.meshResources.length).toBe(2);
    const [a, b] = scene.sceneHierarchy.children;
    expect(a.meshResourceId).not.toBe(b.meshResourceId);

    // and the two variants really do carry different colours
    const colorOf = (id: string | undefined) => {
      const res = scene.meshResources.find((r) => r.id === id)!;
      const mat = scene.gltfMaterials[res.primitives[0].materialIndex] as any;
      return mat.pbrMetallicRoughness.baseColorFactor;
    };
    expect(colorOf(a.meshResourceId)).not.toEqual(colorOf(b.meshResourceId));
  });

  it('includes the layer fallback colour in resource identity', () => {
    const parsed = makeParsed({
      defs: new Map<number | string, ParsedDefinition>([[1, panelDefinition()]]),
      rootInstances: [
        { refIdx: 1, name: 'OnA', matrix: translation(0, 0, 0), layerId: 10 },
        { refIdx: 1, name: 'OnB', matrix: translation(50, 0, 0), layerId: 20 },
      ],
      layerIdToName: new Map([
        [10, 'LayerA'],
        [20, 'LayerB'],
      ]),
      layerColors: new Map<string, [number, number, number]>([
        ['Layer0', [255, 255, 255]],
        ['LayerA', [255, 0, 0]],
        ['LayerB', [0, 0, 255]],
      ]),
    });
    const scene = buildInstancedSceneFromParsed(parsed);

    // The face is unpainted, so the layer colour is what it renders as -
    // two layers with different colours cannot share one resource.
    expect(scene.meshResources.length).toBe(2);
    const [a, b] = scene.sceneHierarchy.children;
    expect(a.meshResourceId).not.toBe(b.meshResourceId);
    expect(a.layer).toBe('LayerA');
    expect(b.layer).toBe('LayerB');
  });

  it('shares one resource for two layers that happen to share a colour', () => {
    const parsed = makeParsed({
      defs: new Map<number | string, ParsedDefinition>([[1, panelDefinition()]]),
      rootInstances: [
        { refIdx: 1, name: 'OnA', matrix: translation(0, 0, 0), layerId: 10 },
        { refIdx: 1, name: 'OnB', matrix: translation(50, 0, 0), layerId: 20 },
      ],
      layerIdToName: new Map([
        [10, 'LayerA'],
        [20, 'LayerB'],
      ]),
      layerColors: new Map<string, [number, number, number]>([
        ['Layer0', [255, 255, 255]],
        ['LayerA', [7, 7, 7]],
        ['LayerB', [7, 7, 7]],
      ]),
    });
    const scene = buildInstancedSceneFromParsed(parsed);
    expect(scene.meshResources.length).toBe(1);
  });

  it('separates variants whose textures differ but average to one colour', () => {
    const sameColor = { r: 141, g: 141, b: 141, a: 255 };
    const parsed = makeParsed({
      defs: new Map<number | string, ParsedDefinition>([[1, panelDefinition()]]),
      rootInstances: [
        { refIdx: 1, name: 'A', matrix: translation(0, 0, 0), materialId: 1 },
        { refIdx: 1, name: 'B', matrix: translation(50, 0, 0), materialId: 2 },
      ],
      materials: new Map([
        ['FabricA', texturedMaterial('FabricA', sameColor, [0xff, 0xd8, 0xff, 0xaa])],
        ['FabricB', texturedMaterial('FabricB', sameColor, [0xff, 0xd8, 0xff, 0xbb, 0xbb])],
      ]),
      materialIdToName: new Map([
        [1, 'FabricA'],
        [2, 'FabricB'],
      ]),
    });
    const scene = buildInstancedSceneFromParsed(parsed);

    expect(scene.meshResources.length).toBe(2);
    expect(scene.textures.length).toBe(2);
  });
});

describe('buildInstancedScene: parity with the baked path', () => {
  /** Same parsed input through both builders must describe the same
   * world-space triangles with the same materials. */
  const expectParity = (parsed: ReturnType<typeof makeParsed>) => {
    const baked = buildSceneFromParsed(parsed);
    const instanced = buildInstancedSceneFromParsed(parsed);

    const bakedTris = flattenBakedScene(baked);
    const instancedTris = flattenInstancedScene(instanced);
    expect(instancedTris.length).toBe(bakedTris.length);

    // Compared in walk order; materials by CONTENT, not index, since the
    // two paths allocate into the same table in their own encounter order.
    const { worstDelta, firstMismatch, materialMismatches } = compareTrianglesInOrder(
      instancedTris,
      bakedTris,
      { actual: instanced.gltfMaterials, expected: baked.gltfMaterials }
    );
    expect(firstMismatch).toBeNull();
    expect(materialMismatches).toBe(0);
    expect(worstDelta).toBeLessThan(1e-5);
    return { baked, instanced };
  };

  it('matches for a simple translated placement', () => {
    expectParity(
      makeParsed({
        defs: new Map<number | string, ParsedDefinition>([[1, panelDefinition()]]),
        rootInstances: [{ refIdx: 1, matrix: translation(10, 20, 30) }],
      })
    );
  });

  it('matches for rotation', () => {
    expectParity(
      makeParsed({
        defs: new Map<number | string, ParsedDefinition>([[1, panelDefinition()]]),
        rootInstances: [{ refIdx: 1, matrix: rotationZ(37, [5, -3, 2]) }],
      })
    );
  });

  it('matches for uniform and non-uniform scale', () => {
    expectParity(
      makeParsed({
        defs: new Map<number | string, ParsedDefinition>([[1, panelDefinition()]]),
        rootInstances: [
          { refIdx: 1, name: 'uniform', matrix: scaleMatrix(2, 2, 2) },
          { refIdx: 1, name: 'nonuniform', matrix: scaleMatrix(3, 0.5, 1.75, [100, 0, 0]) },
        ],
      })
    );
  });

  it('matches for a MIRRORED (negative-determinant) placement', () => {
    expectParity(
      makeParsed({
        defs: new Map<number | string, ParsedDefinition>([[1, panelDefinition()]]),
        rootInstances: [
          { refIdx: 1, name: 'normal', matrix: translation(0, 0, 0) },
          { refIdx: 1, name: 'mirrored', matrix: scaleMatrix(-1, 1, 1, [200, 0, 0]) },
        ],
      })
    );
  });

  it('matches for NESTED instance transforms', () => {
    // outer -> inner -> panel, each level carrying its own transform
    const inner = new GeometryBuilder();
    inner.instances.push({
      offset: 0,
      refGuid: '1',
      refIdx: 1,
      name: 'panel-in-inner',
      matrix: translation(3, 4, 5),
      materialId: null,
      children: [],
    });
    const innerDef: ParsedDefinition = {
      guid: 'INNER', name: 'Inner', isImage: false, alwaysFacesCamera: false, builder: inner,
    };

    const outer = new GeometryBuilder();
    outer.instances.push({
      offset: 0,
      refGuid: '2',
      refIdx: 2,
      name: 'inner-in-outer',
      matrix: rotationZ(90, [10, 0, 0]),
      materialId: null,
      children: [],
    });
    const outerDef: ParsedDefinition = {
      guid: 'OUTER', name: 'Outer', isImage: false, alwaysFacesCamera: false, builder: outer,
    };

    const { instanced } = expectParity(
      makeParsed({
        defs: new Map<number | string, ParsedDefinition>([
          [1, panelDefinition()],
          [2, innerDef],
          [3, outerDef],
        ]),
        rootInstances: [{ refIdx: 3, name: 'outer', matrix: scaleMatrix(2, 2, 2, [7, 8, 9]) }],
      })
    );

    // and the nesting is genuinely preserved, not flattened
    expect(instanced.sceneHierarchy.children.length).toBe(1);
    expect(instanced.sceneHierarchy.children[0].children.length).toBe(1);
    expect(instanced.sceneHierarchy.children[0].children[0].children.length).toBe(1);
  });

  it('matches for a face with a HOLE (multi-loop triangulation)', () => {
    expectParity(
      makeParsed({
        defs: new Map<number | string, ParsedDefinition>([[1, panelWithHoleDefinition()]]),
        rootInstances: [{ refIdx: 1, matrix: translation(1, 2, 3) }],
      })
    );
  });

  it('matches for different FRONT and BACK materials', () => {
    const parsed = makeParsed({
      defs: new Map<number | string, ParsedDefinition>([
        [1, panelDefinition(24, 'TwoSided', { materialId: 1, backMaterialId: 2 })],
      ]),
      rootInstances: [{ refIdx: 1, matrix: translation(0, 0, 0) }],
      materials: new Map([
        ['Red', plainMaterial('Red', { r: 200, g: 10, b: 10, a: 255 })],
        ['Blue', plainMaterial('Blue', { r: 10, g: 10, b: 200, a: 255 })],
      ]),
      materialIdToName: new Map([
        [1, 'Red'],
        [2, 'Blue'],
      ]),
    });
    const { instanced } = expectParity(parsed);

    // front/back genuinely differ, so the definition splits into two
    // single-sided primitives rather than one double-sided one
    expect(instanced.meshResources[0].primitives.length).toBe(2);
    for (const prim of instanced.meshResources[0].primitives) {
      const mat = instanced.gltfMaterials[prim.materialIndex] as any;
      expect(mat.doubleSided).toBeUndefined();
    }
  });

  it('marks a face whose sides resolve alike as doubleSided, once', () => {
    const parsed = makeParsed({
      defs: new Map<number | string, ParsedDefinition>([[1, panelDefinition()]]),
      rootInstances: [{ refIdx: 1, matrix: translation(0, 0, 0) }],
    });
    const { instanced } = expectParity(parsed);
    expect(instanced.meshResources[0].primitives.length).toBe(1);
    const mat = instanced.gltfMaterials[instanced.meshResources[0].primitives[0].materialIndex] as any;
    expect(mat.doubleSided).toBe(true);
  });

  it('retains ROOT loose geometry', () => {
    const parsed = makeParsed({
      defs: new Map<number | string, ParsedDefinition>([[1, panelDefinition()]]),
      rootGeometry: panelDefinition(36, 'LooseFloor'),
      rootInstances: [{ refIdx: 1, matrix: translation(60, 0, 0) }],
    });
    const { instanced } = expectParity(parsed);

    // the root node itself owns a resource for the loose geometry
    expect(instanced.sceneHierarchy.meshResourceId).toBeDefined();
    const rootRes = instanced.meshResources.find(
      (r) => r.id === instanced.sceneHierarchy.meshResourceId
    )!;
    expect(rootRes.definitionId).toBe('ROOT');
  });

  it('matches for many mixed placements at once', () => {
    expectParity(
      makeParsed({
        defs: new Map<number | string, ParsedDefinition>([
          [1, panelDefinition()],
          [2, panelWithHoleDefinition()],
        ]),
        rootGeometry: panelDefinition(30, 'Loose'),
        rootInstances: [
          { refIdx: 1, name: 'a', matrix: translation(0, 0, 0) },
          { refIdx: 1, name: 'b', matrix: rotationZ(45, [40, 0, 0]) },
          { refIdx: 2, name: 'c', matrix: scaleMatrix(1.5, 2.5, 1, [0, 40, 0]) },
          { refIdx: 2, name: 'd', matrix: scaleMatrix(-2, 1, 1, [0, 80, 0]) },
        ],
      })
    );
  });
});

describe('buildInstancedScene: UVs, textures and metadata', () => {
  it('preserves UVs from the inherited textured material', () => {
    const parsed = makeParsed({
      defs: new Map<number | string, ParsedDefinition>([[1, panelDefinition()]]),
      rootInstances: [{ refIdx: 1, matrix: translation(0, 0, 0), materialId: 7 }],
      materials: new Map([['Decor', texturedMaterial('Decor')]]),
      materialIdToName: new Map([[7, 'Decor']]),
    });

    const baked = buildSceneFromParsed(parsed);
    const instanced = buildInstancedSceneFromParsed(parsed);

    const bakedUvs = Array.from(baked.glbPrimitives[0].uvs).sort((a, b) => a - b);
    const instUvs = Array.from(instanced.meshResources[0].primitives[0].uvs).sort((a, b) => a - b);

    // UVs are computed in local space, so an instance transform must not
    // change them at all - the two paths agree exactly.
    expect(instUvs.length).toBe(bakedUvs.length);
    for (let i = 0; i < instUvs.length; i++) {
      expect(instUvs[i]).toBeCloseTo(bakedUvs[i], 6);
    }

    // 24" panel, 24x12" tile: one repeat across, two down.
    const maxU = Math.max(...instUvs.filter((_, i) => i % 1 === 0));
    expect(maxU).toBeGreaterThan(0);
    expect(instanced.textures.length).toBe(1);
  });

  it('deduplicates textures shared by several materials', () => {
    const shared = [0xff, 0xd8, 0xff, 0x42];
    const parsed = makeParsed({
      defs: new Map<number | string, ParsedDefinition>([[1, panelDefinition()]]),
      rootInstances: [
        { refIdx: 1, name: 'A', matrix: translation(0, 0, 0), materialId: 1 },
        { refIdx: 1, name: 'B', matrix: translation(40, 0, 0), materialId: 2 },
      ],
      materials: new Map([
        ['MatA', texturedMaterial('MatA', { r: 5, g: 5, b: 5, a: 255 }, shared)],
        ['MatB', texturedMaterial('MatB', { r: 9, g: 9, b: 9, a: 255 }, shared)],
      ]),
      materialIdToName: new Map([
        [1, 'MatA'],
        [2, 'MatB'],
      ]),
    });
    const scene = buildInstancedSceneFromParsed(parsed);
    expect(scene.textures.length).toBe(1);
  });

  it('reports the same node metadata the baked path reports', () => {
    const parsed = makeParsed({
      defs: new Map<number | string, ParsedDefinition>([[1, panelDefinition()]]),
      rootInstances: [{ refIdx: 1, name: 'Chair', matrix: translation(10, 20, 30), layerId: 5 }],
      layerIdToName: new Map([[5, 'Furniture']]),
      layerColors: new Map<string, [number, number, number]>([
        ['Layer0', [255, 255, 255]],
        ['Furniture', [200, 100, 50]],
      ]),
    });

    const baked = buildSceneFromParsed(parsed);
    const instanced = buildInstancedSceneFromParsed(parsed);

    const b = baked.sceneHierarchy.children[0];
    const i = instanced.sceneHierarchy.children[0];

    expect(i.name).toBe(b.name);
    expect(i.definitionName).toBe(b.definitionName);
    expect(i.layer).toBe(b.layer);
    expect(i.positionMm).toEqual(b.positionMm);
    expect(i.properties).toEqual(b.properties);
  });

  it('rejects a self-instancing definition, like the baked path', () => {
    const selfRef = new GeometryBuilder();
    selfRef.instances.push({
      offset: 0,
      refGuid: '1',
      refIdx: 1,
      name: 'me-inside-me',
      matrix: translation(1, 0, 0),
      materialId: null,
      children: [],
    });
    const parsed = makeParsed({
      defs: new Map<number | string, ParsedDefinition>([
        [1, { guid: 'SELF', name: 'Self', isImage: false, alwaysFacesCamera: false, builder: selfRef }],
      ]),
      rootInstances: [{ refIdx: 1, matrix: translation(0, 0, 0) }],
    });

    expect(() => buildInstancedSceneFromParsed(parsed)).toThrow(/Recursive component definition/);
  });
});

File: packages/cpp/tests/instanced_scene_test.cpp
#include <algorithm>
#include <array>
#include <cmath>
#include <functional>
#include <gtest/gtest.h>
#include <map>
#include <string>
#include <vector>

#include <openskp/openskp.hpp>

#include "test_helpers.hpp"

// Instanced scene building (openskp#200, ported from TypeScript's
// buildInstancedScene()/toInstancedGLB()).
//
// The strongest correctness evidence available: run BOTH builders over the
// repository's real .skp fixtures and require that flattening the
// instanced result reproduces the baked builder's world-space triangles
// exactly. This covers, on genuine files, everything a synthetic test
// would cover piecewise - nested groups/components, instance-painted
// materials, layers, front/back materials, textures, holes, mirrored
// transforms - because whatever those files happen to contain has to come
// out the same either way.

namespace openskp {
namespace {

// Float32 round-off only, same tolerance and justification as the
// TypeScript reference: the baked path transforms in float64 then stores
// the world-space result as float32; the instanced path stores the
// local-space value as float32 and transforms afterwards. Both are
// single-rounding-step correct, but round at different moments, so a
// coordinate can land one float32 ulp apart between them.
constexpr double kTolerance = 1e-5;

std::array<double, 16> mul4(const std::array<double, 16>& a, const std::array<double, 16>& b) {
  std::array<double, 16> out{};
  for (int col = 0; col < 4; ++col) {
    for (int row = 0; row < 4; ++row) {
      double s = 0;
      for (int k = 0; k < 4; ++k) s += a[k * 4 + row] * b[col * 4 + k];
      out[col * 4 + row] = s;
    }
  }
  return out;
}

std::array<double, 3> apply_matrix(const std::array<double, 16>& m, const std::array<float, 3>& p) {
  return {
      m[0] * p[0] + m[4] * p[1] + m[8] * p[2] + m[12],
      m[1] * p[0] + m[5] * p[1] + m[9] * p[2] + m[13],
      m[2] * p[0] + m[6] * p[1] + m[10] * p[2] + m[14],
  };
}

constexpr std::array<double, 16> kIdentity4{
    1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1,
};

struct FlatTri {
  std::array<double, 3> a, b, c;
  std::size_t material_index;
};

// Walk the instanced tree, composing node transforms, and emit every
// triangle in world space - i.e. reconstruct what build_scene() bakes.
// Test-only: the whole point of the instanced output is to avoid
// materialising this.
std::vector<FlatTri> flatten_instanced(const InstancedScene& scene) {
  std::map<std::string, const InstancedMeshResource*> by_id;
  for (auto& r : scene.mesh_resources) by_id[r.id] = &r;

  std::vector<FlatTri> out;
  std::function<void(const InstancedNode&, const std::array<double, 16>&)> visit;
  visit = [&](const InstancedNode& node, const std::array<double, 16>& parent) {
    const auto world = mul4(parent, node.matrix);
    if (node.mesh_resource_id) {
      auto found = by_id.find(*node.mesh_resource_id);
      if (found != by_id.end()) {
        for (auto& prim : found->second->primitives) {
          for (std::size_t i = 0; i + 2 < prim.indices.size(); i += 3) {
            std::array<std::array<double, 3>, 3> tri{};
            for (int k = 0; k < 3; ++k) {
              const auto vi = prim.indices[i + static_cast<std::size_t>(k)];
              tri[static_cast<std::size_t>(k)] = apply_matrix(
                  world,
                  {prim.positions[vi * 3], prim.positions[vi * 3 + 1], prim.positions[vi * 3 + 2]});
            }
            out.push_back(FlatTri{tri[0], tri[1], tri[2], prim.material_index});
          }
        }
      }
    }
    for (auto& child : node.children) visit(child, world);
  };
  visit(scene.scene_hierarchy, kIdentity4);
  return out;
}

std::vector<FlatTri> flatten_baked(const Scene& scene) {
  std::vector<FlatTri> out;
  for (auto& prim : scene.glb_primitives) {
    for (std::size_t i = 0; i + 2 < prim.indices.size(); i += 3) {
      std::array<std::array<double, 3>, 3> tri{};
      for (int k = 0; k < 3; ++k) {
        const auto vi = prim.indices[i + static_cast<std::size_t>(k)];
        tri[static_cast<std::size_t>(k)] = {prim.positions[vi * 3], prim.positions[vi * 3 + 1],
                                            prim.positions[vi * 3 + 2]};
      }
      out.push_back(FlatTri{tri[0], tri[1], tri[2], prim.material_index});
    }
  }
  return out;
}

std::size_t instanced_buffer_bytes(const InstancedScene& scene) {
  std::size_t total = 0;
  for (auto& r : scene.mesh_resources) {
    for (auto& p : r.primitives) {
      total +=
          p.positions.size() * 4 + p.normals.size() * 4 + p.uvs.size() * 4 + p.indices.size() * 4;
    }
  }
  return total;
}

std::size_t baked_buffer_bytes(const Scene& scene) {
  std::size_t total = 0;
  for (auto& p : scene.glb_primitives) {
    total +=
        p.positions.size() * 4 + p.normals.size() * 4 + p.uvs.size() * 4 + p.indices.size() * 4;
  }
  return total;
}

bool materials_equal(const GltfMaterial& a, const GltfMaterial& b) {
  const auto& pa = a.pbr_metallic_roughness;
  const auto& pb = b.pbr_metallic_roughness;
  if (pa.base_color_factor != pb.base_color_factor) return false;
  if (pa.metallic_factor != pb.metallic_factor) return false;
  if (pa.roughness_factor != pb.roughness_factor) return false;
  if (pa.base_color_texture.has_value() != pb.base_color_texture.has_value()) return false;
  if (a.double_sided != b.double_sided) return false;
  return true;
}

void walk_metadata_parity(const InstanceNode& baked, const InstancedNode& instanced) {
  EXPECT_EQ(instanced.name, baked.name);
  EXPECT_EQ(instanced.definition_name, baked.definition_name);
  EXPECT_EQ(instanced.layer, baked.layer);
  EXPECT_EQ(instanced.position_mm, baked.position_mm);
  EXPECT_EQ(instanced.properties, baked.properties);
  ASSERT_EQ(instanced.children.size(), baked.children.size());
  for (std::size_t k = 0; k < baked.children.size(); ++k) {
    walk_metadata_parity(baked.children[k], instanced.children[k]);
  }
}

const std::vector<std::string> kFixtures = {
    "SU_File.skp",
    "Untitled.skp",
    "capilla_quiroz_v17.skp",
    "gondola_v20.skp",
    "single_material_v17.skp",
};

class InstancedSceneParity : public ::testing::TestWithParam<std::string> {};

TEST_P(InstancedSceneParity, ReproducesBuildScenesWorldSpaceTriangles) {
  const auto path = test::fixture(GetParam().c_str());
  const auto baked = SkpFile::open(path).build_scene();
  const auto instanced = SkpFile::open(path).build_instanced_scene();

  const auto baked_tris = flatten_baked(baked);
  const auto instanced_tris = flatten_instanced(instanced);

  ASSERT_EQ(instanced_tris.size(), baked_tris.size());

  double worst_delta = 0;
  std::size_t material_mismatches = 0;
  for (std::size_t i = 0; i < baked_tris.size(); ++i) {
    const auto& a = instanced_tris[i];
    const auto& e = baked_tris[i];
    if (!materials_equal(instanced.gltf_materials[a.material_index],
                         baked.gltf_materials[e.material_index])) {
      ++material_mismatches;
    }

    // The baked path flips a mirrored instance's triangle winding
    // (swapping indices[1]/[2]) so a naive renderer sees correct-facing
    // geometry even after the per-instance transform is baked away and
    // "erased." The instanced path doesn't need that: a mirrored
    // instance's negative-determinant matrix stays on the node, and a
    // spec-compliant glTF consumer flips the front-face convention from
    // that - the standard way glTF represents mirrored placements of
    // shared geometry, which a baked, never-shared primitive has no
    // matrix left to do. So the same triangle can legitimately reach here
    // in either winding; try both pairings and keep whichever is closer.
    const auto delta_for = [](const std::array<double, 3>& p1, const std::array<double, 3>& p2) {
      double d = 0;
      for (int k = 0; k < 3; ++k) {
        d = std::max(d,
                     std::abs(p1[static_cast<std::size_t>(k)] - p2[static_cast<std::size_t>(k)]));
      }
      return d;
    };
    const double direct = std::max({delta_for(a.a, e.a), delta_for(a.b, e.b), delta_for(a.c, e.c)});
    const double reflected =
        std::max({delta_for(a.a, e.a), delta_for(a.b, e.c), delta_for(a.c, e.b)});
    worst_delta = std::max(worst_delta, std::min(direct, reflected));
  }

  EXPECT_EQ(material_mismatches, 0u);
  EXPECT_LT(worst_delta, kTolerance);
}

TEST_P(InstancedSceneParity, NeverStoresMoreGeometryThanTheBakedPath) {
  const auto path = test::fixture(GetParam().c_str());
  const auto baked_bytes = baked_buffer_bytes(SkpFile::open(path).build_scene());
  const auto instanced_bytes = instanced_buffer_bytes(SkpFile::open(path).build_instanced_scene());
  // Equal when nothing repeats; strictly smaller once anything does.
  EXPECT_LE(instanced_bytes, baked_bytes);
}

TEST_P(InstancedSceneParity, ResolvesTheSameLayersAndDynamicPropertiesPerNode) {
  const auto path = test::fixture(GetParam().c_str());
  const auto baked = SkpFile::open(path).build_scene();
  const auto instanced = SkpFile::open(path).build_instanced_scene();
  walk_metadata_parity(baked.scene_hierarchy, instanced.scene_hierarchy);
}

INSTANTIATE_TEST_SUITE_P(RealFixtures, InstancedSceneParity, ::testing::ValuesIn(kFixtures));

}  // namespace
}  // namespace openskp
